vm/starnix/starnix.go

```diff
@@ -0,0 +1,395 @@
+	for i := 0; ; i++ {
+		inst, err := pool.ctor(workdir, index)
+		if err == nil {
+			return inst, nil
+		}
+		return nil, err
+	}
+}
+
+func (pool *Pool) ctor(workdir string, index int) (vmimpl.Instance, error) {
+	inst := &instance{
+		fuchsia:  pool.cfg.Fuchsia,
+		name:     fmt.Sprintf("VM-%v", index),
+		index:    index,
+		cfg:      pool.cfg,
+		debug:    pool.env.Debug,
+		timeouts: pool.env.Timeouts,
+		workdir:  workdir,
+		executor: filepath.Join(workdir, "adb_executor.sh"),
+	}
+	closeInst := inst
+	defer func() {
+		if closeInst != nil {
+			closeInst.Close()
+		}
+	}()
+
+	var err error
+	inst.rpipe, inst.wpipe, err = osutil.LongPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := inst.boot(); err != nil {
+		return nil, err
+	}
+
+	closeInst = nil
+	return inst, nil
+}
+func (inst *instance) adbPids() ([]string, error) {
+	cmd := osutil.Command("lsof", "-ti", fmt.Sprintf(":%d", inst.port))
+	rawOutput, err := cmd.Output()
+	if err != nil {
+		return nil, err
+	}
+	pidsString := strings.Trim(string(rawOutput), "\n ")
+	return strings.Split(pidsString, "\n"), nil
+}
+
+func (inst *instance) terminateFuchsiaVm() {
+	if inst.debug {
+		log.Logf(0, "Terminating Fuchsia VM: %v", inst.name)
+	}
+	inst.runCommand("ffx", "emu", "stop", inst.name)
+}
+
+func (inst *instance) Close() {
+	if inst.starnix != nil {
+		inst.terminateFuchsiaVm()
+	}
+	if inst.merger != nil {
+		inst.merger.Wait()
+	}
+	if inst.rpipe != nil {
+		inst.rpipe.Close()
+	}
+	if inst.wpipe != nil {
+		inst.wpipe.Close()
+	}
+}
+
+func (inst *instance) startFuchsiaVm() error {
+	cmd, err := inst.runCommand("ffx", "emu", "start", "--headless", "--name", inst.name)
+	if err != nil {
+		return err
+	}
+	inst.starnix = cmd
+	return nil
+}
+
+func (inst *instance) startFuchsiaLogs() error {
+	cmd := osutil.Command("ffx", "--target", inst.name, "log")
+	cmd.Dir = inst.fuchsia
+	cmd.Stdout = inst.wpipe
+	cmd.Stderr = inst.wpipe
+	inst.merger.Add("fuchsia", inst.rpipe)
+	return cmd.Start()
+}
+
+func (inst *instance) startAdbServerAndConnection(timeout time.Duration) error {
+	cmd := osutil.Command("ffx", "--target", inst.name, "starnix", "adb", "-p", fmt.Sprintf("%d", inst.port))
+	cmd.Dir = inst.fuchsia
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	inst.adb = cmd
+	if err := inst.waitForAdb(timeout); err != nil {
+		return err
+	}
+	return inst.connectToAdb(timeout)
+}
+
+func (inst *instance) connectToAdb(timeout time.Duration) error {
+	startTime := time.Now()
+	for {
+		if inst.debug {
+			log.Logf(0, "Attempting to connect to ADB")
+		}
+		cmd := osutil.Command("adb", "connect", fmt.Sprintf("127.0.0.1:%d", inst.port))
+		connect_output, err := cmd.Output()
+		if err == nil && strings.HasPrefix(string(connect_output), "connected to") {
+			return nil
+		}
+		cmd = osutil.Command("adb", "disconnect", fmt.Sprintf("127.0.0.1:%d", inst.port))
+		cmd.Run()
+		if inst.debug {
+			log.Logf(0, "adb connect failed")
```

Comment:
Nit: "disconnect failed"?

Reply:
The thing is, if `adb connect` fails you need to do the disconnection in order to try connecting again. I don't really know what to do if `adb disconnect` fails. This part of the code is based on this [script](https://cs.opensource.google/fuchsia/fuchsia/+/main:src/proc/tests/android/adb/adb_art_test.py;l=41-46).

Reply:
Oh, yes, I misread! I thought this was checking the output of `adb disconnect`. That makes sense, nevermind!